src/queue.zig

```diff
@@ -3,126 +3,220 @@ const assert = std.debug.assert;
 
 const constants = @import("./constants.zig");
 
+const QueueLink = extern struct {
+    next: ?*QueueLink = null,
+};
+
 /// An intrusive first in/first out linked list.
 /// The element type T must have a field called "next" of type ?*T
 pub fn QueueType(comptime T: type) type {
     return struct {
-        const Queue = @This();
-
-        in: ?*T = null,
-        out: ?*T = null,
-        count: u64 = 0,
+        any: QueueAny,
 
-        // This should only be null if you're sure we'll never want to monitor `count`.
-        name: ?[]const u8,
+        pub const Link = QueueLink;
+        const Queue = @This();
 
-        // If the number of elements is large, the constants.verify check in push() can be too
-        // expensive. Allow the user to gate it. Could also be a comptime param?
-        verify_push: bool = true,
+        pub inline fn init(options: struct {
+            name: ?[]const u8,
+            verify_push: bool = true,
+        }) Queue {
+            return .{ .any = .{
+                .name = options.name,
+                .verify_push = options.verify_push,
+            } };
+        }
 
-        pub fn push(self: *Queue, elem: *T) void {
-            if (constants.verify and self.verify_push) assert(!self.contains(elem));
+        pub inline fn push(self: *Queue, link: *T) void {
+            self.any.push(&link.link);
+        }
 
-            assert(elem.next == null);
-            if (self.in) |in| {
-                in.next = elem;
-                self.in = elem;
-            } else {
-                assert(self.out == null);
-                self.in = elem;
-                self.out = elem;
-            }
-            self.count += 1;
+        pub inline fn pop(self: *Queue) ?*T {
```

Comment:
This is just a hunch, but I suspect that _inline_ here can _reduce_ compilation time, as it should lead to less code arriving at LLVM.

Reply:
Zig’s rule of thumb for `inline fn` is either to propagate `comptime` arguments or because a benchmark showed it’s worth using. However, we often see `inline fn` used in Zig’s standard library for simple dispatches like this one!

So, I think we can consider it a good practice :+1: